runtime/compiler/x/codegen/J9TreeEvaluator.cpp

```diff
@@ -10806,6 +10806,172 @@ inlineMathSQRT(
 
    }
 
+static bool inlineMathFMAFloat(TR::Node *node, TR::CodeGenerator *cg)
+   {
+   // fma(a, b, c) => (a*b+c)
+   TR::Node *firstChild = node->getChild(0);
+   TR::Node *secondChild = node->getChild(1);
+   TR::Node *thirdChild = node->getChild(2);
+   
+   TR::Register *targetRegister = NULL;
+
+   if(firstChild->getOpCode().isLoadConst() && 
+      secondChild->getOpCode().isLoadConst() && 
+      thirdChild->getOpCode().isLoadConst())
+      {
+      union
+         {
+         int32_t bits;
+         float value;
+         } a, b, c, d;
+      // d = a*b + c
+      a.value = firstChild->getFloat();
+      b.value = secondChild->getFloat();
+      c.value = thirdChild->getFloat();
+
+      bool aIsNan = ((a.bits & 0x7f800000) == 0x7f800000) && ((a.bits & 0x007fffff) != 0);
+      bool bIsNan = ((b.bits & 0x7f800000) == 0x7f800000) && ((b.bits & 0x007fffff) != 0);
+      bool cIsNan = ((c.bits & 0x7f800000) == 0x7f800000) && ((c.bits & 0x007fffff) != 0);
+      bool aIsInf = ((a.bits & 0x7fffffff) == 0x7f800000);
+      bool bIsInf = ((b.bits & 0x7fffffff) == 0x7f800000);
+      bool cIsPosInf = ((c.bits & 0xffffffff) == 0x7f800000);
+      bool cIsNegInf = ((c.bits & 0xffffffff) == 0xff800000);
+      bool aIsZero = (a.bits & 0x7fffffff) == 0;
+      bool bIsZero = (b.bits & 0x7fffffff) == 0;
+
+      if(aIsNan || bIsNan || cIsNan ||
+        (aIsInf && bIsZero) || (aIsZero && bIsInf))
+         {
+         // result is NaN
+         d.bits = 0x7ff00000;
+         }
+      else if((aIsInf && !bIsZero && !bIsNan) ||
+              (bIsInf && !aIsZero && !aIsNan))
+         {
+         bool mulIsPositive = ((a.bits & 0x80000000) ^ (b.bits & 0x80000000) == 0);
+         if((mulIsPositive && cIsNegInf) || (!mulIsPositive && cIsPosInf))
+            {
+            // result is NaN
+            d.bits = 0x7ff00000;
+            }
+         else 
+            {
+            d.value = TR::Compiler->arith.floatAddFloat(TR::Compiler->arith.floatMultiplyFloat(a.value, b.value), c.value);
+            }
+         }
+      else
+         {
+         d.value = TR::Compiler->arith.floatAddFloat(TR::Compiler->arith.floatMultiplyFloat(a.value, b.value), c.value);
+         }
+      
+      TR::IA32ConstantDataSnippet *cds = cg->findOrCreate4ByteConstant(firstChild, d.bits);
+
+      targetRegister = cg->allocateRegister(TR_FPR);
+      generateRegMemInstruction(MOVSSRegMem, node, targetRegister, generateX86MemoryReference(cds, cg), cg);
+      }
+   else 
+      {
+      TR::Register *src1Register = cg->evaluate(firstChild);
+      TR::Register *src2Register = cg->evaluate(secondChild);
+      TR::Register *src3Register = cg->evaluate(secondChild);
+
+      targetRegister = src1Register;
+
+      generateRegRegRegInstruction(VFMADD132SSRegRegReg, node, targetRegister, src3Register, src2Register, cg);    
```

Comment:
You should also try and produce the best possible instruction sequence given the context in which this operation occurs.  For instance, in the case of the `vfmadd` instruction there are a number of variants to this instruction that you should apply depending on whether one of its children's registers dies at this instruction or if it is the first and only use of the child.

Looking at what has been added to the instruction list it appears you should add the following to `X86Ops.ins`:
```
VFMADD132RegRegReg (already done)
VFMADD132RegRegMem
VFMADD213RegRegReg
VFMADD213RegRegMem
VFMADD231RegRegReg
VFMADD231RegRegMem
```
Of the 3 children to this `Math.fma()` call, if the reference count on any of those child nodes is 1 prior to their evaluation (i.e., `child->getReferenceCount() == 1 && child->getRegister() != NULL`) then that means that it is the last reference to this child node and any register it is evaluated into dies after this instruction.  If there is such a child then you want that one to be in the `xmm1` position for these instructions because it will always be clobbered by the instruction.  Ultimately this will save some register shuffling.

Furthermore, if a child's reference count is 1 and it has never been evaluated into a register before (i.e., `child->getReferenceCount() == 1 && child->getRegister() == NULL`) then that means this is the first and only evaluation of that child.  In such a case, you want to have this child in the `xmm3` position so that you can use the memory form of the instruction.  Floating point multiplication is commutative, so you should be able to cover all the permutations.  This will save a register because doing it otherwise the child would need to be loaded from memory into a register (potentially evicting something) when its not really required.

In any case, choose the variant of the instruction that allows you to achieve that.  Remember, the order of the operands on the instruction does not have to match the order of the children on the call.